### ISSUE TITLE
Can't Assign Inputs with Xbox 360 Controller on Windows 10 Pro 2004

### ISSUE THREAD

**Issue by IcyGirl75** [project64/project64#1774]
Hello,

I hope this is the right place to put this issue, as it's about the N-Rage Input V2 plugin.

When I open the window to assign buttons with my Xbox 360 controller, when I click any of the buttons and then press a button on my controller what it shows is "Keyboard ???" instead of "Gamepad: Button 02" for example.

Occasionally when trying to assign inputs, the very first input will sometimes show correctly, but then afterwards all I get is "Keyboard ???" as soon as I click any of the buttons in the configuration window. This doesn't stop happening until I close and re-open the emulator.

This issue has been happening specifically on Windows 10 2004 for me. It happens on both my upgraded install from Windows 10 1909 and it also happened on a completely fresh install of Windows 10 2004. I had no problems before upgrading to Windows 10 2004, it worked on my Windows 10 1909 setup.

If I enable XInput, I can configure the plugin normally, so the problem is only when that option is deselected. Configuring the plugin to use keyboard keys also still works. The problem seems to be specifically with controllers with XInput disabled. I could only test with Xbox 360 controllers.

This problem is present in version 2.5.3.1196-g05f9 of the plugin, but it also happens with the version that comes with version 2.3 of Project64.

Thanks!
![000349 - N-Rage Input V2 - Configuration (Small)](https://user-images.githubusercontent.com/66268364/83434103-143ccd00-a3f7-11ea-8f9c-fbbb83e1d250.jpg)

**Comment by Skullnick** [comment]
Same prob!

**Comment by DaRkL3AD3R** [comment]
Yep exact same problem and it's even happening with a very old build of the N-Rage plugin I used to have. What I want to know is why my plugin shows version number in chinese symbols. The whole thing is super sketchy.

**Comment by achraf-boussaada** [comment]
same problem here. The same controller used to work fine couple of months ago. I reinstalled project 64 today probably an other version of what I had before and now I can't map my controller

**Comment by ClaudioMuller** [comment]
Hey there,
I currently have the exact same issue, except with a PS4 controller.
It worked without a hitch until a couple of days ago. When I started the emulator today, all of a sudden it didn't work anymore.

As described above:
When I now go into the "Configure Controller Plugin" menu, and press to any of the keys I want to assign, I will get the "keyboard: ???" message and am unable to assign any key. Neither on my controller, nor on my keyboard.

I have uninstalled Project 64 and re-installed it. If I then tried to configure the controls, it worked as long as I used the keyboard. As soon as I plug in my PS4-controller and try to assign a button, I will get the "keyboard: ???" message. At that point, it will do the same, even if I try to input a key from my keyboard. What might also be worth mentioning is this: If I now press the up- and down-key on my keyboard, the blue border around the button I tried to assign will toggle up and down. It's almost as if at that point, the menu thinks I've switched to an option where I want to use my keyboard to toggle through the menu, instead of assigning keys to emulator-buttons.

When I disconnect the controller and re-start the emulator, I'm able again to configure the controls with my keyboard, until I'm trying to use the controler again.

As I said: I was able to use the emulator with my PS4-controller until a few days ago without a problem. This is entirely new, but I cannot get it to work anymore. I tried to installe an older version of Project 64, however that same problem happens with that as well.

**Comment by project64** [comment]
does project-input work for you? https://www.pj64-emu.com/nightly-builds

**Comment by IcyGirl75** [comment]
Yes, for the most part, the new plugin is working for me!

I just tested it and I did however run into some issues. I tested Project64 2.4.0.1364-gf4106c9 and the version of Project64 Input Plugin that came with it, version 1.0.0.1364-gf4106c9.

There are three separate issues that I spotted:

It is possible to get "Keyboard: ???" to show up even in this plugin. All I had to do to see this in the configuration window was to press "Setup" to assign all the buttons and then move the analog sticks around rapidly. It shows up quite frequently. It can be seen on "R-Trigger" in the picture.

![000372 - Configure Input](https://user-images.githubusercontent.com/66268364/89261316-14219f00-d5eb-11ea-8281-bb15f481e42a.jpg)

The second issue that I spotted was that some of the axis's are reversed from how they should be. I think they show up properly in the configuration window, but upon testing them in GoldenEye 007, they appeared to be reversed. This could be seen from watching the controller in the game's pause menu as I pressed buttons. The reversed axis's are: x and y rotation, and z axis. The axis's for the left analog stick were correct. This happens with all controllers 1-4.

And the final issue I encountered was that after assigning a controller's buttons with this plugin, in order to assign a different controller's buttons to the already assigned buttons on a different controller, I had to press "Reset", otherwise the buttons would still only work from the controller they were originally assigned with.

For example, I assign all the buttons on the "Player 1" tab to Xbox 360 controller 1 and save the configuration. I then reassign all the button assignments on the "Player 1" tab with Xbox 360 controller 2 and save the configuration. I proceed to try the buttons in a game and find they still work on controller 1 and don't work on controller 2. Only "Reset", followed by reassigning the buttons with the controller I want to use seems to fix this. Closing and re-opening the emulator didn't help.

Other than that, the plugin seemed to work great! No crashes at all while I was testing, and my Xbox One Elite controllers worked as well, albeit with the same issues as the 360 controllers.

Thanks for your reply!

**Comment by DaRkL3AD3R** [comment]
Things are in a really bad state, I don't know what went wrong. I feel like I was better off on my very old build 2.3.2 from the site, using a custom N-Rage plugin I had from like 2008 that had Xinput support modded into it. I can't find it anymore, and I feel like it's never gonna go back to the compatibility level it was at then. And then there's the GlideN64 compilation situation where I don't even know where to get the latest builds anymore. Man what a mess.

**Comment by project64** [comment]
in general the zip version from the nightly builds (https://www.pj64-emu.com/nightly-builds) should have the latest GlideN64 version there. It should be compiling any new changes and have what ever the latest version that was compiled at the time of making that version of pj64

**Comment by rrodg84** [comment]
i had the same problem. my workaround was to buy a logitech f310 gamepad for 15 bucks on amazon. it has aswitch on the back to go from xinput to dinput for legacy games. i snagged an older version of nrage dinput plugin from the ultrahle emulator that i found on emulator zone. it works fine, i just have to switch to dinput to play n64 games and back to xinput for literally everything else on my computer. i tried a bunch of different version of nrage xinput and other input plugs and only one that i can get to work is nrage directx 8 input plugin v1.61. none of the xinput plugs will recognize my xinput gamepad (the original xbox 360 controller or the logitech switched to xinput mode) and just does the ??? keyboard thing. i never tried any of the nightly builds or any of that stuff

**Comment by zmarotrix** [comment]
Same issue using a Direct input controller. This issue has persisted for me for years across multiple computers.

**Comment by Squall-Leonhart** [comment]
this has been fixed
